Availability sets older duplicates to Cancelled, fixes #372

abrophy opened the pull request:
This method naively compares start and end times.

I can have it round to the closest 1/4 hour but I wasn't sure if we have an abstraction for rounding of time we're using in the project.

Can easily implement that in the method, or just do the rounding in the method itself, whatever you prefer.

abrophy committed:
Availability sets older duplicates status to Cancelled on create, fixes #372

kgf commented:
I can't look right now, but I believe we do have code for rounding to the nearest quarter hour.

Sadly, the tests did not pass.

abrophy commented:
I'll get on this right away.

kgf commented on app/models/availability.rb:
```diff
@@ -1,6 +1,8 @@
 class Availability < ActiveRecord::Base
   belongs_to :person
 
+  before_create :cancel_duplicates
```
I think this would be better as an after save hook. We would want it on save, so that if someone updated an availability to overlap, we'd still catch that. And an after hook, so that if the availability _doesn't_ save for some reason, we won't have cancelled availabilities we should not have.

abrophy replied:
I agree with this, will do.

kgf commented on spec/models/availability_spec.rb:
```diff
@@ -34,4 +34,16 @@
+    it 'sets status for previous availabilities with matching start and end times to Cancelled' do
+      start_time = Time.now
+      end_time = start_time + 2.minutes
+      first_availability = create(:availability, person: a_person, start_time: start_time, end_time: end_time, created_at: Time.now - 2.minutes)
```
Sorry, I'm not following this. Why are you setting the created_at ?

abrophy replied:
Is unnecessary, you're right.

kgf commented on spec/models/availability_spec.rb:
```diff
@@ -34,4 +34,16 @@
+      first_availability = create(:availability, person: a_person, start_time: start_time, end_time: end_time, created_at: Time.now - 2.minutes)
+      second_availability = create(:availability, person: a_person, start_time: start_time, end_time: end_time)
+
+      expect(first_availability.reload.status).to eq('Cancelled')
```
Needing the reload here makes sense to me. Line 42 saves a record into the database, and loads a copy into memory, referenced by _first_availability_. Your hook then changes the underlying persisted record, but does not reload the object.
You could `first_availability.reload`, then `expect(first_availability.status).to eq("Cancelled")`

abrophy replied:
I'll give this a try

abrophy committed:
corrects test for Availability.cancel_duplicates

No longer creates availability with explicit created_at, and moves reloading of model
into its own expression to avoid race conditions

abrophy committed:
cancel_duplicate method doesn't change status of self

abrophy committed:
Availability cancels duplicates on after_create hook

moves from after_create callback from before_create callback

abrophy commented on spec/models/availability_spec.rb:
```diff
@@ -39,10 +39,11 @@
       second_availability = create(:availability, person: a_person, start_time: start_time, end_time: end_time)
 
-      expect(first_availability.reload.status).to eq('Cancelled')
+      first_availability = first_availability.reload
```
moving the reload into its own expression like this still hasn't fixed the issue.
Test is read on Travis but green on local machine.

pcai commented on app/models/availability.rb:
```diff
@@ -39,4 +41,15 @@ def self.process_data
+    previous_availabilities = person.availabilities.for_time_span(start_time..end_time).order(:created_at)
+    previous_availabilities.each do |a|
+      if start_time == a.start_time && end_time == a.end_time
+        a.update(status: "Cancelled") unless self == a
```
Is there a reason you are using `for_time_span(start_time..end_time)` instead of just using an equality search to get precisely the records that need to be cancelled? e.g., something like `person.availabilities.where(start_time: self.start_time, end_time: self.end_time).where.not(id: self.id)`

I think that way you can omit `unless self == a` and the time comparison, since you'll avoid loading the `self` record and anything that doesnt match.

kgf replied:
This would be a fine first approach. Right now folks have no way to correct a mistake without involving someone who has access to the web UI. The most common use case is an availability for a specific event, so the times would exactly match.

pcai replied:
Right. If that's the case, my suggestion is to replace this whole block with something like: 
```ruby
# cancel all availabilities that are exact duplicates of this one
person.availabilities.where(start_time: self.start_time, end_time: self.end_time)
  .where.not(id: self.id).update_all(status: "Cancelled")
```

I haven't tested this but I think that captures the essence of what this is trying to accomplish?

pcai replied:
Sorry, just saw the linked issue suggests the other approach. I'll take a look to see if I can shed any light on the travis failure

mxie replied:
I agree with what @pcai has said about rewriting the block. If we want to use `for_time_span`, I think it'd look something like this:

```ruby
person.availabilities.for_time_span(start_time..end_time).where.not(id: id).update_all(status: "Cancelled")
```

pcai commented:
The tests are probably failing on travis for one of 2 reasons:
* the time resolution is not correct. saving and retrieving from the database might cause the equality comparison to fail.
* a timezone issue is causing the save and retrieval to make the equality comparison fail.

I'm not familiar enough with travis to know if there's an ssh/debug option but you could always add a debug commit that ups the verbosity to confirm or deny

mxie commented on spec/models/availability_spec.rb:
```diff
@@ -34,4 +34,17 @@
   end
+
+  context 'cancel_duplicates' do
+    it 'sets status for previous availabilities with matching start and end times to Cancelled' do
```
Would you mind rearranging the phrasing of this to say "sets status to X for..."? It took me a few reads before realizing what it was saying. 😬

mxie commented on spec/models/availability_spec.rb:
```diff
@@ -34,4 +34,17 @@
+      first_availability = create(:availability, person: a_person, start_time: start_time, end_time: end_time)
+      second_availability = create(:availability, person: a_person, start_time: start_time, end_time: end_time)
+
+      first_availability = first_availability.reload
```
Do we have to reload?

mxie replied:
Nevermind - I see why now. :)

mxie commented on spec/models/availability_spec.rb:
```diff
@@ -34,4 +34,17 @@
+
+  context 'cancel_duplicates' do
+    it 'sets status for previous availabilities with matching start and end times to Cancelled' do
+      start_time = Time.now
```
We should use `Time.current` instead: https://robots.thoughtbot.com/its-about-time-zones

kgf commented:
@abrophy I was able to get this to reliably pass by making these changes.
https://github.com/ReadyResponder/ReadyResponder/pull/399/files/8b9f45c76e6c6ff8c198e2534d0103d56fd86e4a..e7ab945191e710092de1f3fa35b3e78d160b0e05
I doubt the removal of the `order` clause was important, but it was there so....
Would you like to make the changes and update this PR, or perhaps you're busy and would prefer that I merge #399 .
I'd like to cut a release soon, so just let me know. 
Thanks for your help !
Kevin

abrophy commented:
Sorry I had other projects come up in the mean time,  thanks to everyone who looked over this
